- TF-IDF is used because it helps us identify the most important words in a collection of text documents (like articles, emails, or books) by finding which words stand out

In [7]:
from sklearn.feature_extraction.text import TfidfVectorizer

corpus=[
    "Thor is eating pizza , Loki is eating pizza , Ironman  ate pizza already",
    "Apple is announcing new Iphone Tomorrow",
    "Tesla is announcing new model-3 tomorrow",
    "google is announcing new pixel-6 tomorrow",
    "Microsoft is abbouncing new surface tomoroow",
    "Amazon is announcing new eco-dot tomorrow",
    "I am eating biryani and you are eating grapes"
]


In [8]:
v = TfidfVectorizer()

# Fit the vectorizer on the corpus and transform the data
v.fit(corpus)
transform_output = v.transform(corpus)

# Print the vocabulary generated by the vectorizer
print(v.vocabulary_)

{'thor': 26, 'is': 17, 'eating': 11, 'pizza': 23, 'loki': 18, 'ironman': 16, 'ate': 8, 'already': 1, 'apple': 6, 'announcing': 5, 'new': 21, 'iphone': 15, 'tomorrow': 28, 'tesla': 25, 'model': 20, 'google': 13, 'pixel': 22, 'microsoft': 19, 'abbouncing': 0, 'surface': 24, 'tomoroow': 27, 'amazon': 3, 'eco': 12, 'dot': 10, 'am': 2, 'biryani': 9, 'and': 4, 'you': 29, 'are': 7, 'grapes': 14}


In [10]:
#print idf of each word
all_feature_names=v.get_feature_names_out()
for word in all_feature_names:
    indx=v.vocabulary_.get(word)
    idf_score=v.idf_[indx]
    print(f"{word}:{idf_score}")

abbouncing:2.386294361119891
already:2.386294361119891
am:2.386294361119891
amazon:2.386294361119891
and:2.386294361119891
announcing:1.4700036292457357
apple:2.386294361119891
are:2.386294361119891
ate:2.386294361119891
biryani:2.386294361119891
dot:2.386294361119891
eating:1.9808292530117262
eco:2.386294361119891
google:2.386294361119891
grapes:2.386294361119891
iphone:2.386294361119891
ironman:2.386294361119891
is:1.1335313926245225
loki:2.386294361119891
microsoft:2.386294361119891
model:2.386294361119891
new:1.2876820724517808
pixel:2.386294361119891
pizza:2.386294361119891
surface:2.386294361119891
tesla:2.386294361119891
thor:2.386294361119891
tomoroow:2.386294361119891
tomorrow:1.4700036292457357
you:2.386294361119891


In [11]:
import pandas as pd

df=pd.read_csv("Ecommerce_data.csv")
print("shape:",df.shape)
df.head(5)

shape: (24000, 2)


Text                   label
0  Urban Ladder Eisner Low Back Study-Office Comp...               Household
1  Contrast living Wooden Decorative Box,Painted ...               Household
2  IO Crest SY-PCI40010 PCI RAID Host Controller ...             Electronics
3  ISAKAA Baby Socks from Just Born to 8 Years- P...  Clothing & Accessories
4  Indira Designer Women's Art Mysore Silk Saree ...  Clothing & Accessories

In [12]:
df['label'].value_counts()

label
Household                 6000
Electronics               6000
Clothing & Accessories    6000
Books                     6000
Name: count, dtype: int64

#IMP: balanced data and imbalance data
we can see that there are equal no.of times and perfectly almost all labels occurred
There is no problem of class imbalnace and hence no need to apply balancing technique

In [27]:
label_mapping = {
    'Household': 0,
    'Books': 1,
    'Electronics': 2,
    'Clothing & Accessories': 3  # Adjusted to match the exact label names
}

# Map the labels to numeric values
df['label_num'] = df['label'].map(label_mapping)

# Display the DataFrame to confirm mapping
df.head(5)


Text                   label  \
0  Urban Ladder Eisner Low Back Study-Office Comp...               Household   
1  Contrast living Wooden Decorative Box,Painted ...               Household   
2  IO Crest SY-PCI40010 PCI RAID Host Controller ...             Electronics   
3  ISAKAA Baby Socks from Just Born to 8 Years- P...  Clothing & Accessories   
4  Indira Designer Women's Art Mysore Silk Saree ...  Clothing & Accessories   

   label_num  
0          0  
1          0  
2          2  
3          3  
4          3

Train Test split


In [33]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    df.Text,
    df.label_num,
    test_size=0.2,
    random_state=2022,
    stratify=df.label_num
)


In [35]:

print(X_train.shape)

print(X_test.shape)


(19200,)
(4800,)


In [37]:
X_train.head()

15820    IRIS Furniture Children Deluxe Spiderman Toddl...
23224    Godox CB-09 Hard Carrying Storage Suitcase Car...
4638     Ugreen All in 1 USB 3.0 Card Reader USB Memory...
15245    Spread Spain Metallic Gold Bar Trolley/Kitchen...
5378     Chromozome Men's Calf Socks (Pack of 3) (SX-3 ...
Name: Text, dtype: object

In [56]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report
from sklearn.feature_extraction.text import TfidfVectorizer

clf=Pipeline([
    ('vectorizer_tfidf',TfidfVectorizer()),
    ('KNN',KNeighborsClassifier())
])

clf.fit(X_train,y_train)

y_pred=clf.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.95      0.96      0.95      1200
           1       0.97      0.95      0.96      1200
           2       0.97      0.97      0.97      1200
           3       0.97      0.98      0.97      1200

    accuracy                           0.96      4800
   macro avg       0.96      0.96      0.96      4800
weighted avg       0.96      0.96      0.96      4800



In [60]:
y_test[:5]

20706    0
19166    2
15209    3
2462     1
6621     3
Name: label_num, dtype: int64

In [62]:
y_pred[:5]

array([0, 2, 3, 1, 0], dtype=int64)

#There are classes which misclassified so 4% are loss in accuracy

actual_classes=3 
but predicted =0
______________________________
20706    0
19166    2
15209    3
2462     1
6621     **3**
_______________________________
array([0, 2, 3, 1, **0**]